proposals/0127-get-sysvar-syscall.md

```diff
@@ -0,0 +1,233 @@
+swell the syscall interface as changes are made, making maintainence and
+reimplementation difficult.
+
+## Alternatives Considered
+
+Other mechanisms considered that provide sysvar data to BPF programs are as
+follows:
+
+### Sysvar-Specific Syscalls
+
+This approach would involve changing nothing. We'd continue on the same course,
+directly coupling new sysvar fetch functions to new syscalls.
+
+### New VM Memory Regions
+
+A new memory region in the VM could be introduced to store sysvar data, which
+could then be made available to BPF programs through an interface around raw
+pointers, which loads the data similarly to the interface proposed herein.
+
+This approach would depend on direct mapping of host memory to VM memory,
+which has been in development for some time and requires further research
+and development to introduce safely to the VM ABI. This extensive R&D period
+would mean Core BPF program initiatives would remain blocked.
+
+Without direct mapping, sysvar data would be fully copied to this memory
+region each time a VM is created (currently one per transaction processed),
+causing a degradation in runtime performance and likely reducing compute
+by far less compared to the compute savings posed by this proposal.
+
+Even with direct mapping, this change makes address translation more complex
+and therefore slows down _all_ memory accesses.
+
+## New Terminology
+
+N/A.
+
+## Detailed Design
+
+A single `sol_get_sysvar` syscall can be defined, which can be used by all
+forward-facing sysvar interfaces to retrieve data from any sysvar.
+
+```c
+/**
+ * Retrieves a slice of data from a sysvar and copies it to VM memory.
+ *
+ * @param sysvar_id The identifier of the sysvar to retrieve data from.
+ * @param var_addr  VM memory address to copy the retrieved data to.
+ * @param offset    The offset to start copying data from.
+ * @param length    The length of data to copy, in bytes.
+ * @return          A 64-bit unsigned integer error code:
+ *                    - 0 if the operation is successful.
+ *                    - Non-zero error code.
+ *
+ * If the operation is not successful, data will not be written to the
+ * provided VM memory address.
+ */
+uint64_t sol_get_sysvar(
+  /* r1 */ void const * sysvar_id,
+  /* r2 */ uint8_t *    var_addr,
+  /* r3 */ uint64_t     offset,
+  /* r4 */ uint64_t     length,
+);
+
+```
+
+### Control Flow
+
+The syscall aborts the virtual machine if any of these conditions are true:
+
+- Not all bytes in VM memory range `[sysvar_id, sysvar_id + 32)` are readable.
+- Not all bytes in VM memory range `[var_addr, var_addr + length)` are writable.
+- `offset + length` is not in `[0, 2^64)`.
+- `var_addr + length` is not in `[0, 2^64)`.
+- Compute budget is exceeded.
+
+The syscall returns `1` if the sysvar is not present in the Sysvar Cache.
+
+The syscall returns `2` if `offset + length` is greater than the size of the
+sysvar.
```

Comment:
if we are introducing a distinction between hard and soft errors (successful failures?) we need to specify what order the checks happen in, because a set of inputs that could trigger both a soft and a hard error needs to trigger the same one on all impls

im inclined to want to check the _soft_ conditions first, and bail with the appropriate code if they are met, because there is no possibility of the language runtime accidentally returning them later

the more i think about this tho the more worried i get about the possibility of an attacker crafting a transaction where writing to the output buffer triggers a segfault on one validator impl but not the other

Reply:
incidentally im assuming this means that a `length` of 0 should noop and return `Ok(0)`

Reply:
The intention by @ripatel-fd was to have "graceful errors" that won't abort the VM, but maybe there's a case to be made to get rid of these. I believe the intention was to handle a scenario where an on-chain program may use the syscall directly (without the handler) and give them the ability to gracefully handle errors, rather than the VM panicking.

https://github.com/solana-foundation/solana-improvement-documents/pull/127#discussion_r1544351726

However, I'm pretty sure we dove into the error propagation after this discussion, and if a syscall throws an `InstructionError`, that will bubble up to a `ProgramError` in the executing program, so perhaps _all_ of these syscall errors are actually graceful?

https://discord.com/channels/428295358100013066/967028962746327060/1222861983842570294

Reply:
oh to be clear my preferred solution would be to redefine the nonzero return cases to also abort, unless theres a concrete usecase we are hoping to support with this

Reply:
Right I agree on condensing the control flow, but I'm saying I don't think it actually aborts. I think the error bubbles up. It would have to panic to abort.

Reply:
as far as i can tell, returning an error does abort unconditionally, after testing with `Clock::get` this afternoon and again now. this must happen somewhere inside solana-rbpf but because of that fact plus ffi, macros, the separation between syscall registration and invocation, and thiserror, it is too difficult for me to untangle and locate the place where this actually occurs

what i do know is `solana_program::syscalls::sol_get_clock_sysvar` returns a `u64` rather than a `Result`, and control never resumes inside the solana-program `Sysvar::get` impl if you return an `Err` inside `SyscallGetClockSysvar`

@Lichtso could you enlighten us?

Reply:
So, @2501babe you're correct that syscall errors abort the VM. I just ran a little test to investigate further. Turns out I already ran this test, lol.

Here's the block of code in question.

https://github.com/anza-xyz/agave/blob/297a7aa40a42d5dad4002d11da10ef3748876041/program-runtime/src/invoke_context.rs#L514-L534

```rust
let result = match vm.program_result {
    ProgramResult::Ok(_) => {
        stable_log::program_success(&logger, &program_id);
        Ok(())
    }
    ProgramResult::Err(ref err) => {
        if let EbpfError::SyscallError(syscall_error) = err {
            if let Some(instruction_err) = syscall_error.downcast_ref::<InstructionError>()
            {
                stable_log::program_failure(&logger, &program_id, instruction_err);
                Err(instruction_err.clone())
            } else {
                stable_log::program_failure(&logger, &program_id, syscall_error);
                Err(InstructionError::ProgramFailedToComplete)
            }
        } else {
            stable_log::program_failure(&logger, &program_id, err);
            Err(InstructionError::ProgramFailedToComplete)
        }
    }
};
```

The `vm.program_result` is a `StableResult<u64, EbpfError>`. As you can see from the snippet, the program's result can be any `Ok(u64)`, but we bury that `u64` value. We only care that it was successful.

The snippet also shows that valid `InstructionError` values can be bubbled up, although **the VM is still aborted**.

It's not possible to catch a syscall error, even if it is a valid `InstructionError`.

```rust
fn process_instruction(
    _program_id: &Pubkey,
    accounts: &[AccountInfo],
    _instruction_data: &[u8],
) -> ProgramResult {
    let mut accounts_iter = accounts.iter();

    let vote_account = next_account_info(&mut accounts_iter)?;

    msg!("Log!");

    if let Err(e) = get_epoch_stake(vote_account.key) {
        msg!("Error: {:?}", e);
    }

    msg!("Another log!");

    // Should always return Ok
    Ok(())
}
```

The reason the syscall definitions and the program stubs define the syscall as returning a `u64`, is because it _may_ return an `InstructionError` that can be bubbled up to the client, such as a `BanksClient` or `RpcClient`, but the VM has aborted nonetheless.

I guess it means we can use `InstructionError` to define specific error codes from within the syscall, if we want. Anything that throws an `EbfError`, such as [`translate`](https://github.com/anza-xyz/agave/blob/033bcfcab5696839c0ad10486ddf4c85fc73358b/programs/bpf_loader/src/syscalls/mod.rs#L480-L490) - even if it's an `EbfError::SyscallError`, is going to route to `InstructionError::ProgramFailedToComplete`.

Reply:
With _all of that_ being said, @ripatel-fd was making the case of gracefully returning an error, which can only be done by doing `Ok(1)`, `Ok(2)`, etc.

With that in mind, ordering shouldn't _really_ matter, since anything that throws within the syscall is automatically an abort, while returned `Ok(..)` values are returned. The handler is then responsible for translating those `Ok(T)` values into `Ok(bytes)` or `Err(ProgramError::V)`.

Reply:
In that spirit, I think the specification is accurate.

Reply:
returning `Ok(1)` doesnt halt execution tho. so if there is a case where one impl hits an `Err` whereas the other hits one of the `Ok` cases first, there would be a consensus failure. the order has to be de defined or else the `Ok` errors have to change to `Err`

Reply:
Ahhh, I see what you're saying. I think it probably makes sense to say if you provide a condition that should abort the VM, then the VM should be aborted, period.

In other words, VM violations should take precedence. One should only receive a graceful error if the VM wouldn't be violated otherwise. Open to discussion that, though.

Reply:
If you hate what I just added, then I can do pseudocode, but it seemed sufficient and concise.

Reply:
We were planning to merge today, and looks like based on above there's acceptance of what I suggested, then committed?
@2501babe @ripatel-fd

Reply:
think we got ourselves a simd!